src/shell/hwloc.c

```diff
@@ -0,0 +1,100 @@
+/************************************************************\
+ * Copyright 2024 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* hwloc options handler
+ */
+#define FLUX_SHELL_PLUGIN_NAME "hwloc"
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <flux/core.h>
+#include <flux/shell.h>
+
+#include "src/common/libutil/read_all.h"
+#include "ccan/str/str.h"
+
+#include "builtins.h"
+
+static int create_xmlfile (flux_shell_t *shell)
+{
+    int fd = -1;
+    char *xmlfile = NULL;
+    const char *hwloc_xml;
+    const char *tmpdir;
+
+    if (!(tmpdir = flux_shell_getenv (shell, "FLUX_JOB_TMPDIR")))
+        tmpdir = "/tmp";
+
+    if (flux_shell_get_hwloc_xml (shell, &hwloc_xml) < 0) {
+        shell_log_error ("failed to get shell hwloc xml");
+        goto error;
+    }
+    if (asprintf (&xmlfile, "%s/hwloc.xml", tmpdir) < 0) {
+        shell_log_error ("asprintf HWLOC_XMLFILE failed");
+        goto error;
+    }
+    if ((fd = open (xmlfile, O_CREAT|O_EXCL|O_WRONLY, 0640)) < 0) {
+        shell_log_errno ("%s", xmlfile);
+        goto error;
+    }
+    shell_debug ("Writing %ld bytes to HWLOC_XMLFILE=%s\n",
+                 strlen (hwloc_xml),
+                 xmlfile);
+    if (write_all (fd, hwloc_xml, strlen (hwloc_xml)) < 0) {
+        shell_log_errno ("failed to write HWLOC_XMLFILE");
+        goto error;
+    }
+    close (fd);
```

Comment:
Probably should check the return value of close since it can fail on a write.

Reply:
I also fixed the batch plugin which also failed to check the return code of `close(2)` (though only when using the old style batch jobspec options)